docs(k8s): update docs (#2229)
scaleway/docs-content: containers/kubernetes/reference-content/secure-cluster-with-private-network.mdx

Scaleway Kubernetes Kapsule provides a managed environment to create, configure and run a cluster of preconfigured machines for containerized applications. This allows you to create Kubernetes clusters without the complexity of managing the infrastructure.

All new Kubernetes clusters are deployed with a Scaleway [Private Network](/network/vpc/concepts/#private-networks) using [controlled isolation](#what-is-the-difference-between-controlled-isolation-and-full-isolation).

<Message type="tip">
  We recommend [migrating any existing clusters](#how-can-i-migrate-my-existing-clusters-to-private-networks) to a Private Network via the Scaleway console before Q1/2024, as **public-only endpoints will become deprecated.**
</Message>

By default, worker nodes are currently delivered with public IP addresses ([controlled isolation](#what-is-the-difference-between-controlled-isolation-and-full-isolation)). These IPs serve solely for outgoing traffic from your nodes to the internet; no default services are set to listen on them.

> Even though these nodes have public IP addresses for specific maintenance and operational purposes, your cluster's security remains uncompromised. See below for more information. Optionally you can configure your nodes inside a 100% private network using [full isolation](#what-is-the-difference-between-controlled-isolation-and-full-isolation)

Reviewer: ```suggestion
Even though these nodes have public IP addresses for specific maintenance and operational purposes, your cluster's security remains uncompromised. See below for more information. Optionally, you can configure your nodes inside a 100% private network using [full isolation](#what-is-the-difference-between-controlled-isolation-and-full-isolation)
```

* Less manual configuration work such as security group configuration, IP range configuration etc.
* Lower latency

> You can use the Scaleway console to both create new Private Networks, and add clusters to existing Private Networks. Each cluster is autoconfigured using a /22 IP subnet thanks to Scaleway’s IPAM.

Reviewer: ```suggestion
You can use the Scaleway console to create new Private Networks and add clusters to existing Private Networks. Each cluster is autoconfigured using a /22 IP subnet, thanks to Scaleway’s IPAM.
```

## Can I use a Public Gateway with my Private Network to exit all outgoing traffic from the nodes?

> Yes, you can use a Public Gateway to exit outgoing traffic from your Kapsule nodes having private IPs. For more information, how to configure a Public Gateway, refer to our [Public Gateway documentation](/network/public-gateways/how-to/configure-a-public-gateway/).

Reviewer: ```suggestion
Yes, you can use a Public Gateway to exit outgoing traffic from your Kapsule nodes having private IPs. For more information on how to configure a Public Gateway, refer to our [Public Gateway documentation](/network/public-gateways/how-to/configure-a-public-gateway/).
```

## Scaleway product compatibility

### Can I use a Public Gateway with my Private Network to exit all outgoing traffic from the nodes?

Yes, you are required to attach a Private Gateway when setting up a node pool with [full isolation](#what-is-the-difference-between-controlled-isolation-and-full-isolation). This allows Kapsule nodes with private IPs to route their outgoing traffic through the Public Gateway. For detailed steps on setting up a Public Gateway, refer to our [Public Gateway documentation](/network/public-gateways/how-to/configure-a-public-gateway/).
Keep in mind that removing or detaching the Public Gateway from the Private Network can cause a single point of failure in the cluster, preventing fully isolated node pools from accessing the control plane.

### Is Kosmos compatible with Private Networks?

Only Kapsule can use a Private Network.

Kosmos uses Kilo as a CNI, which uses WireGuard to create a VPN Mesh between nodes for communication between pods. Any node in Kosmos, either in Scaleway or outside, uses these VPN tunnels to communicate securely by construct.

### Are Managed Databases compatible with Kubernetes Kapsule on Private Networks?

Yes they are.

> Since July 2023, the automatic allocation of IP addresses via IPAM is available for Managed Databases. These IP addresses are compatible with Scaleway's VPC, which is now in General Availability and can be used with, not only Managed Databases, but all products in the Scaleway ecosystem.

Reviewer 2: Not quite, there are some incompatible products (e.g. Messaging and Queuing for one). Maybe link to this page https://www.scaleway.com/en/docs/network/vpc/troubleshooting/vpc-limitations/

Author: ```suggestion
Since July 2023, the automatic allocation of IP addresses via IPAM is available for Managed Databases. These IP addresses are compatible with Scaleway's VPC, which is now in General Availability. For more information about product compatibility refer to the [VPC documentation]([/network/vpc/troubleshooting/vpc-limitations/](https://www.scaleway.com/en/docs/network/vpc/troubleshooting/vpc-limitations/).
```

For any new Private Networks you create and attach to Managed Databases after July 2023, your private IP addresses are automatically allocated.

> If you have set up Private Network endpoints for your Managed Databases prior to July 2023, and want to connect to Kapsule via a Private Network, you must first delete your old private network endpoint. Then, you can create a new one, either via the [Scaleway console](/managed-databases/postgresql-and-mysql/how-to/connect-database-private-network/#how-to-attach-a-database-instance-to-a-private-network) or API.

Reviewer 2: In the console would this be the equivalent of detaching and reattaching the Private Network?

In the example below we show you how to do so via the API. We specify the automated configuration of your Private Network via IPAM using `"ipam_config": {},`.

```bash
curl --request POST \
  --url https://api.scaleway.com/rdb/v1/regions/$REGION/instances/$INSTANCE_ID/endpoints \
  --header "Content-Type: application/json" \
  --header "X-Auth-Token: $SCW_SECRET_KEY" \
  --data '{
	"endpoint_spec": {
		"private_network": {
			"ipam_config": {},
			"private_network_id": "<PRIVATE_NETWORK_ID>"
		}
	}
}'
```

<Message type="note">
  Replace `<PRIVATE_NETWORK_ID>` with the ID of the Private Network in question.
</Message>

<Message type="important">
  - This action adds a new endpoint. If you want to use it in your environments, you need to update the endpoint in your configuration.
</Message>

Refer to the [Managed Database for PostgreSQL and MySQL API documentation](https://www.scaleway.com/en/developers/api/managed-database-postgre-mysql/#path-endpoints-create-a-new-database-instance-endpoint) for further information.

> ### Are Managed Load Balancers compatible with Kubernetes Kapsule Private Networks?

Reviewer 2: ```suggestion
### Are managed Load Balancers compatible with Kubernetes Kapsule Private Networks?
```

Managed Load Balancers support Private Networks with private backends and public frontends, meaning the traffic is forwarded to your worker nodes through your clusters' Private Network.

> Additionally, [private Load Balancers](/network/load-balancer/concepts/#private-load-balancer) are supported. These Load Balancers have no public IPs in their back- and frontends.

Reviewer 2: ```suggestion
Additionally, [private Load Balancers](/network/load-balancer/concepts/#private-load-balancer) are supported. These Load Balancers have no public IPs in either their back or frontends.
```

| Description | Worker nodes are assigned both private IPs and public IPs. All inbound traffic on public interface is dropped by default using Security Groups. | Worker nodes are set without public IPs (100% private network). A [Public Gateway](/network/public-gateways/how-to/configure-a-public-gateway/) is required. | Clusters without a Private Network attached. |
| Benefits    | 1. Strong security<br /> 2. Dynamic public IPs to reach out to external providers while avoiding rate limiting                                        | 1. Maximum security <br />2. A stable egress IP for secure connection to external providers         | n/a                                          |
| Notice      | Default choice for new clusters. Can be used in combination with pools using full isolation.                                                    | Requires a Public Gateway, which incurs additional costs.                                     | Deprecated in October 2023.                  |

> ## Can I have clusters with pools using both, controlled isolation and full isolation?

Reviewer: ```suggestion
## Can I have clusters with pools using both controlled and full isolation?
```

> Yes, you can configure clusters with Kapsule pools using controlled isolation and full isolation. Note, that pools using full isolation require a [Public Gateway](/network/public-gateways/how-to/configure-a-public-gateway/) to reach the Kubernetes control plane of the cluster.

Reviewer: ```suggestion
Yes, you can configure clusters with Kapsule pools using controlled isolation and full isolation. Note that pools using full isolation require a [Public Gateway](/network/public-gateways/how-to/configure-a-public-gateway/) to reach the Kubernetes control plane of the cluster.
```

<Message type="important">
  Removing or detaching the Public Gateway from a Kapsule pool with full isolation can lead to a single point of failure in the cluster, as nodes will no longer be able to reach their control plane.
</Message>

## Is Kosmos compatible with Private Networks?

Only Kapsule can use a Private Network.
Kosmos uses Kilo as a CNI, which uses WireGuard to create a VPN Mesh between nodes for communication between pods. Any node in Kosmos, either in Scaleway or outside, uses these VPN tunnels to communicate securely by construct.

## What future options will there be for isolation?

In the future, additional features are planned, such as:

> - **Control Plane in isolation** with nodes and API communicating in the same isolated network. The CNI’s network policies will restrict/allow a range of IPs or ports to control who can access to the API server.

Reviewer 3: ```suggestion
- **Control plane in isolation** with nodes and API communicating in the same isolated network. The CNI’s network policies will restrict/allow a range of IPs or ports to control who can access the API server.
```
